Fixes for Sqlite schema regex

GwendolenLynch opened the pull request:
Both column collation & comments were not respecting word boundaries so `foo` and `foobar` would conflict depending on order in the schema.

Capable & happy to make adjustments to tests, commits, etc, as requested.

GwendolenLynch committed:
[Tests] Check for correct collation when column names are similar

GwendolenLynch committed:
Match column collation based on word boundaries

GwendolenLynch committed:
[Tests] Permutations for Sqlite column comment detection

GwendolenLynch committed:
Match column comments based on word boundaries

rolandow commented:
Could this please be merged?

lcobucci commented:
@GawainLynch please change the destination branch to `master` and rebase your branch to sync with the latest changes we'll handle the backporting.

lcobucci commented on tests/Doctrine/Tests/DBAL/Schema/SqliteSchemaManagerTest.php:
```diff
@@ -41,6 +40,61 @@ public function getDataColumnCollation()
+     */
+    public function testParseColumnCommentFromSQL($comment, $column, $sql)
+    {
+        $conn = $this->getMockBuilder('Doctrine\DBAL\Connection')->disableOriginalConstructor()->getMock();
```
This can be simplified to `$conn = $this->createMock(Connection::class);`

lcobucci commented on tests/Doctrine/Tests/DBAL/Schema/SqliteSchemaManagerTest.php:
```diff
@@ -41,6 +40,61 @@ public function getDataColumnCollation()
+    public function testParseColumnCommentFromSQL($comment, $column, $sql)
+    {
+        $conn = $this->getMockBuilder('Doctrine\DBAL\Connection')->disableOriginalConstructor()->getMock();
+        $conn->expects($this->any())->method('getDatabasePlatform')->will($this->returnValue(new SqlitePlatform()));
```
This can be simplified to `$conn->method('getDatabasePlatform')->willReturn(new SqlitePlatform());`

lcobucci commented on tests/Doctrine/Tests/DBAL/Schema/SqliteSchemaManagerTest.php:
```diff
@@ -41,6 +40,61 @@ public function getDataColumnCollation()
+    /**
+     * @dataProvider getDataColumnComment
+     */
+    public function testParseColumnCommentFromSQL($comment, $column, $sql)
```
Please add the type declarations here (return type as well).

lcobucci commented on tests/Doctrine/Tests/DBAL/Schema/SqliteSchemaManagerTest.php:
```diff
@@ -41,6 +40,61 @@ public function getDataColumnCollation()
+        $ref = new \ReflectionMethod($manager, 'parseColumnCommentFromSQL');
+        $ref->setAccessible(true);
+
+        $this->assertEquals($comment, $ref->invoke($manager, $column, $sql));
```
Please use `self::assert*()` instead (since the method is static)

lcobucci replied:
Also, it looks that `assertSame()` is more appropriated here

lcobucci commented on tests/Doctrine/Tests/DBAL/Schema/SqliteSchemaManagerTest.php:
```diff
@@ -41,6 +40,61 @@ public function getDataColumnCollation()
+
+    public function getDataColumnComment()
+    {
+        return array(
```
Please use short array syntax everywhere 😄

lcobucci reviewed (changes requested):
@GawainLynch thanks for your contribution!

After your rebase you'll see that `./vendor/bin/phpcs` is available, please use it to check the CS violations you have on your PR =)

GawainLynch commented:
Done @lcobucci … FWIW, I agree on all points, I just stuck with what was there initially.

As this is a bugfix, and it affecting numerous users negatively, can we request it is backported to the 2.6 branch for the next patch release, please?

Ocramius commented:
Yes, it will be backported once we have time to work on it.

On 9 Oct 2017 05:13, "Gawain Lynch" <notifications@github.com> wrote:

> Done @lcobucci <https://github.com/lcobucci> … FWIW, I agree on all
> points, I just stuck with what was there initially.
>
> As this is a bugfix, and it affecting numerous users negatively, can we
> request it is backported to the 2.6 branch for the next patch release,
> please?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/doctrine/dbal/pull/2865#issuecomment-335059877>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAJakHSSFsSUnO67nrI6d5ddk7zRddF6ks5sqY9hgaJpZM4PZxrN>
> .
>

k42b3 commented:
It looks like there is still a problem with the regexp. In case we have the following SQL:

```
CREATE TABLE psx_table_command_test (id INTEGER NOT NULL, col_bigint BIGINT NOT NULL, col_binary BLOB NOT NULL, col_blob BLOB NOT NULL, col_boolean BOOLEAN NOT NULL, col_datetime DATETIME NOT NULL, col_datetimetz DATETIME NOT NULL, col_date DATE NOT NULL, col_decimal NUMERIC(10, 0) NOT NULL, col_float DOUBLE PRECISION NOT NULL, col_integer INTEGER NOT NULL, col_smallint SMALLINT NOT NULL, col_text CLOB NOT NULL, col_time TIME NOT NULL, col_string VARCHAR(255) NOT NULL, col_array CLOB NOT NULL --(DC2Type:array)
, col_object CLOB NOT NULL --(DC2Type:object)
, col_json CLOB NOT NULL --(DC2Type:json_array)
, col_guid CHAR(36) NOT NULL --(DC2Type:guid)
, PRIMARY KEY(id))
```

and we try to get the comment for the column "col_decimal" we get `(DC2Type:guid)` as comment, later on this results then in the wrong type. All other columns are working correctly.

lcobucci commented on lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php:
```diff
@@ -463,7 +463,7 @@ private function parseColumnCollationFromSQL($column, $sql)
     {
         if (preg_match(
-            '{[\s(,](?:'.preg_quote($this->_platform->quoteSingleIdentifier($column)).'|'.preg_quote($column).')
+            '{[\s(,](?:\W'.preg_quote($this->_platform->quoteSingleIdentifier($column)).'\W|\W'.preg_quote($column).'\W)
```
As pointed out on https://github.com/doctrine/dbal/pull/2865#issuecomment-341418300 this regex doesn't work for some scenarios, can you please check that?

k42b3 replied:
I have checked the solution but the problem still occurs. You could add the following test case to reproduce the error:

```php
'Column with numeric but no comment' => [
	false, 'a', 'CREATE TABLE "a" ("a" NUMERIC(10, 0) NOT NULL, "b" CLOB NOT NULL --(DC2Type:array)' . "\n" . ', "c" CHAR(36) NOT NULL --(DC2Type:guid)' . "\n" . ')',
],
```

In this case I would assume `false` for column "a" but it returns `(DC2Type:guid)`

lcobucci reviewed (changes requested):
Can you also ensure that the modified lines of `lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php` follows our standards? You can use `vendor/bin/phpcs lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php` for that

GawainLynch commented:
. Can you also ensure that the modified lines of `lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php` follows our standards?

Done @lcobucci 

> I have checked the solution but the problem still occurs.

@k42b3 I think I have it now, that hurt my brain :wink:

k42b3 commented:
@GawainLynch thanks for the fix, this looks good from my POV

GawainLynch commented:
No worries, sorry for going quiet on your comment … it's been one of those Novembers :smile:

lcobucci committed:
Use NULL instead of false when parsing SQL queries

The reasoning behind this is to essentially rely on a single type (or
the absence of it), therefore add type declarations.

Since these are private methods we can do that without breaking BC.

lcobucci commented:
@GawainLynch 🚢  and backported to `v2.6.x` (with some minor changes). Thanks for your contribution!

GawainLynch commented:
Thanks @lcobucci … Appreciated :+1:
